File: alignxy.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import minimize, NonlinearConstraint
from anchormean0to1 import anchormean0to1

    
def alignxy(G_0):

    print("Beginning joint XY alignment")

    padSize=sum(np.isnan(G_0[:,0])) / 2
    
    
    nE = G_0.shape[1]
    nG = G_0.shape[2]
    
    alpha_0 = np.zeros([nG*nE,1]) # One alpha for each profile
    beta_0 = np.ones([nG*nE,1]) # One beta for each profile
    gamma_0 = np.zeros([nE,1]) # One gamma for each embryo
    
    #G_mean_0 = np.nanmean(G_0,1)
    
    nAlpha = np.size(alpha_0)
    nBeta = np.size(beta_0)
    nGamma = np.size(gamma_0)
    
    # Combine parameters into a column array
    p_0 = np.vstack([alpha_0,beta_0,gamma_0])
    
    fun = lambda params: chisq(params, G_0)
    solution = minimize(fun, p_0, method='Nelder-Mead', tol=1e-6)
    
    
def chisq(params, G):
    nP = params.shape[0]
    
    
    # G: 1000 x 102 x 4
    nE = G.shape[1]
    nG = G.shape[2]
    
    alpha = params[0 : nG*nE]
    beta = params[nG*nE : 2*nG*nE]
    gamma = params[2*nG*nE : nP]
    
    iP = 0
    for iE in range(0, nE):
        G[:,iE,:] = np.roll(G[:,iE,:], gamma[iE])
        
    g_mean = np.squeeze(np.nanmean(G,axis=1))
    
    chi2_G = np.zeros([nG,1])
    for iE in range(0, nE):
        for iG in range(0, nG):
            G_tmp = np.squeeze(G[:,:,iG])
            
            # Working on getting param indexing right
            alpha_tmp = alpha[iE*nG - nG : iE*nG]
            beta_tmp = beta[iE*nG - nG : iE*nG]
            
            chi2_G[iG] = np.nanmean(np.nansum( G_tmp - (alpha_tmp + beta_tmp*g_mean[:,iG])**2 ))
            

            
            
    
    #np.roll == circshift
    
    
    
    
    

# =============================================================================
# # alignxy.m:13
#     nAlpha=numel(alpha_0)
# # alignxy.m:15
#     nBeta=numel(beta_0)
# # alignxy.m:16
#     nGamma=numel(gamma_0)
# # alignxy.m:17
#     options=optimoptions('ga','UseParallel',UseParallel_TF,'UseVectorized',false)
# # alignxy.m:19
#     nP=nAlpha + nBeta + nGamma
# # alignxy.m:23
#     ## Specify indices to rereference parameters in cost function.
#     for iG in arange(1,nG).reshape(-1):
#         alphaStart=dot(nAlpha,(iG / nG)) - nE + 1
# # alignxy.m:27
#         alphaEnd=alphaStart + nE - 1
# # alignxy.m:28
#         idcs.alpha[iG]=arange(alphaStart,alphaEnd)
# # alignxy.m:29
#         betaStart=nAlpha + dot(nBeta,(iG / nG)) - nE + 1
# # alignxy.m:31
#         betaEnd=betaStart + nE - 1
# # alignxy.m:32
#         idcs.beta[iG]=arange(betaStart,betaEnd)
# # alignxy.m:33
#     
#     idcs.gamma = copy(arange(nP - nE + 1,nP))
# # alignxy.m:35
#     f=lambda P=None: chisq(P,G_0,idcs)
# # alignxy.m:37
#     IntCon=arange((nP - nGamma + 1),nP)
# # alignxy.m:39
#     lb_alpha=- ones(nAlpha,1)
# # alignxy.m:41
#     ub_alpha=ones(nAlpha,1)
# # alignxy.m:42
#     lb_beta=zeros(nBeta,1)
# # alignxy.m:44
#     ub_beta=dot(2,ones(nBeta,1))
# # alignxy.m:45
#     lb_gamma=repmat(- padSize,nE,1)
# # alignxy.m:47
#     ub_gamma=repmat(padSize,nE,1)
# # alignxy.m:48
#     lb=concat([[lb_alpha],[lb_beta],[lb_gamma]])
# # alignxy.m:50
#     ub=concat([[ub_alpha],[ub_beta],[ub_gamma]])
# # alignxy.m:51
#     P_final,__,exitFlag,__,__=ga(f,nP,[],[],[],[],lb,ub,[],IntCon,options,nargout=5)
# # alignxy.m:53
#     exitFlag
#     ## Extract parameters from P_final.
#     alpha=zeros(nG,nE)
# # alignxy.m:59
#     beta=zeros(nG,nE)
# # alignxy.m:60
#     gamma=zeros(1,nE)
# # alignxy.m:61
#     for iE in arange(1,nE).reshape(-1):
#         for iG in arange(1,nG).reshape(-1):
#             alpha[iG,iE]=P_final(idcs.alpha[iG](iE))
# # alignxy.m:65
#             beta[iG,iE]=P_final(idcs.beta[iG](iE))
# # alignxy.m:66
#         gamma[iE]=P_final(idcs.gamma(iE))
# # alignxy.m:68
#     
#     ## Apply parameters alpha, beta, and gamma to G_0.
#     G=zeros(size(G_0))
# # alignxy.m:72
#     g=zeros(size(G))
# # alignxy.m:73
#     for iE in arange(1,nE).reshape(-1):
#         G[arange(),iE,arange()]=circshift(G_0(arange(),iE,arange()),gamma(iE),1)
# # alignxy.m:75
#     
#     for iG in arange(1,nG).reshape(-1):
#         alpha_tmp=alpha(iG,arange())
# # alignxy.m:78
#         beta_tmp=beta(iG,arange())
# # alignxy.m:79
#         g[arange(),arange(),iG]=(G(arange(),arange(),iG) - alpha_tmp) / beta_tmp
# # alignxy.m:80
#     
#     g,__=anchormean0to1(g,nargout=2)
# # alignxy.m:83
#     shiftAmt=- round((mean(gamma)))
# # alignxy.m:85
#     for iE in arange(1,nE).reshape(-1):
#         g[arange(),iE,arange()]=circshift(g(arange(),iE,arange()),shiftAmt,1)
# # alignxy.m:87
#     
#     toc
#     return g,alpha,beta,gamma
#     
# if __name__ == '__main__':
#     pass
#     
#     ## Cost function
#     
# def chisq(P=None,G=None,idcs=None,*args,**kwargs):
#     varargin = chisq.varargin
#     nargin = chisq.nargin
# 
#     __,nE,nG=size(G,nargout=3)
# # alignxy.m:95
#     alpha=zeros(nG,nE)
# # alignxy.m:97
#     beta=zeros(nG,nE)
# # alignxy.m:98
#     gamma=zeros(1,nE)
# # alignxy.m:99
#     for iE in arange(1,nE).reshape(-1):
#         for iG in arange(1,nG).reshape(-1):
#             alpha[iG,iE]=P(idcs.alpha[iG](iE))
# # alignxy.m:103
#             beta[iG,iE]=P(idcs.beta[iG](iE))
# # alignxy.m:104
#         gamma[iE]=P(idcs.gamma(iE))
# # alignxy.m:106
#         G[arange(),iE,arange()]=circshift(G(arange(),iE,arange()),gamma(iE),1)
# # alignxy.m:107
#     
#     
#     g_mean=squeeze(nanmean(G,2))
# # alignxy.m:110
#     chi2_G=zeros(nG,1)
# # alignxy.m:112
#     for iG in arange(1,nG).reshape(-1):
#         G_tmp=squeeze(G(arange(),arange(),iG))
# # alignxy.m:114
#         alpha_tmp=alpha(iG,arange())
# # alignxy.m:115
#         beta_tmp=beta(iG,arange())
# # alignxy.m:116
#         chi2_G[iG]=nanmean(nansum((G_tmp - (alpha_tmp + multiply(beta_tmp,g_mean(arange(),iG)))) ** 2))
# # alignxy.m:117
#     
#     
#     chi2=mean(chi2_G)
# # alignxy.m:121
#     return chi2
#     
# if __name__ == '__main__':
#     pass
# =============================================================================
    

File: alignxy.m
function [y, alpha, beta, gamma] = alignxy(Y_0, varargin)
%ALIGNXY jointly aligns data along X- and Y- axes of data in a 3D matrix.
% 
% Input:
%   > Y_0: initial profile data. 3D (no. data points) x (no. replicates) x
%   (no. targets).
%   > varargin: NaN pad size as a percentage of the total length. NaN pad
%   is used as a buffer to shift profiles left and right along the X-axis
%   without causing data points to wrap around to the opposite side. If
%   left empty, a default value of 0.05 is used.
% Output:
%   > y: aligned output data. 3D. Same size as input (pads removed).
%   > alpha: additive parameter (1 per profile per target)
%   > beta: scaling/multiplicative parameter (1 per profile per target)
%   > gamma: shifting parameter (1 per replicate, all targets within a
%   replicate are shifted by the same amount)
tic

[nPoints, nReplicates, nTargets] = size(Y_0);

%% Apply NaN padding to either side of each profile.
defaultNanPadSize = 0.05;
switch numel(varargin)
    case 0
        nanPadSize = defaultNanPadSize;
    case 1
        nanPadSize = varargin{1};
end

nNansInPad = round(nPoints * nanPadSize);
Y_0 = padarray(Y_0, nNansInPad, NaN, 'both');

% nNansInPad = sum(isnan(Y_0(:,1,1)))/2;

%% Initialize alignment parameters.
alpha_0 = zeros(nTargets, nReplicates);
beta_0 = ones(nTargets, nReplicates);
gamma_0 = zeros(1, nReplicates);

nAlpha = numel(alpha_0);
nBeta = numel(beta_0);
nGamma = numel(gamma_0);

options = optimoptions('ga', ...
    'UseVectorized', false);

nP = nAlpha + nBeta + nGamma;

%% Specify indices to rereference parameters in cost function.
for iT = 1:nTargets
    alphaStart = nAlpha*(iT/nTargets) - nReplicates + 1;
    alphaEnd = alphaStart + nReplicates - 1;
    idcs.alpha{iT} = alphaStart:alphaEnd;
    
    betaStart = nAlpha + nBeta*(iT/nTargets) - nReplicates + 1;
    betaEnd = betaStart + nReplicates - 1;
    idcs.beta{iT} = betaStart:betaEnd;
end
idcs.gamma = nP - nReplicates + 1: nP;


f = @(P)chisq(P, Y_0, idcs); 

%% Set alignment parameter constraints.
IntCon = (nP - nGamma + 1):nP;

lb_alpha = -ones(nAlpha,1);
ub_alpha = ones(nAlpha,1);

lb_beta = zeros(nBeta,1);
ub_beta = 2*ones(nBeta,1);

lb_gamma = repmat(-nNansInPad/2,nReplicates,1);
ub_gamma = repmat(nNansInPad/2,nReplicates,1);

lb = [lb_alpha; lb_beta; lb_gamma];
ub = [ub_alpha; ub_beta; ub_gamma];

%% Optimization.
% Need to replace GA with something more efficient.
[P_final,~,exitFlag,~,~] = ...
    ga(f,nP,[],[],[],[],lb,ub,[],IntCon,options);

exitFlag

%% Extract parameters from P_final.
alpha = zeros(nTargets,nReplicates);
beta = zeros(nTargets,nReplicates);
gamma = zeros(1,nReplicates);

for iR = 1:nReplicates
    for iT = 1:nTargets
        alpha(iT,iR) = P_final(idcs.alpha{iT}(iR));
        beta(iT,iR) = P_final(idcs.beta{iT}(iR));
    end
    gamma(iR) = P_final(idcs.gamma(iR));
end

%% Apply parameters alpha, beta, and gamma to profile data.
Y = zeros(size(Y_0));
y = zeros(size(Y));
for iR = 1:nReplicates
    Y(:,iR,:) = circshift( Y_0(:,iR,:),gamma(iR),1 );
end
for iT = 1:nTargets
    alpha_tmp = alpha(iT,:);
    beta_tmp = beta(iT,:);
    y(:,:,iT) = (Y(:,:,iT) - alpha_tmp)./beta_tmp;
end

shiftAmt = -round((mean(gamma)));
for iR = 1:nReplicates
    y(:,iR,:) = circshift(y(:,iR,:),shiftAmt,1 );
end

%% Remove NaN padding from either side of each profile.
[nPoints, nReplicates, nTargets] = size(y);

nFrontNans = round( sum(isnan(y(1:round(nPoints/2),:,:)),'all') / ...
    (nReplicates*nTargets) );
nEndNans = round( sum(isnan(y(round(nPoints/2):end,:,:)),'all') / ...
    (nReplicates*nTargets) );

y_temp = y( (nFrontNans : (nPoints - nEndNans + 1)), :, : );

% Ensure the size of the output matches the size of the input.
[n_points_temp,~,~] = size(y_temp);
if n_points_temp == nPoints
    y = y_temp;
elseif n_points_temp == nPoints - 2
    nFrontNans = nFrontNans - 1;
    nEndNans = nEndNans - 1;
    
    y = y( (nFrontNans : (nPoints - nEndNans + 1)), :, : );
elseif n_points_temp == nPoints - 1
    nFrontNans = nFrontNans - 1;
    y = y( (nFrontNans : (nPoints - nEndNans + 1)), :, : );
end

toc

end

%% Cost function
function chi2 = chisq(P, Y, idcs)
    [~,nReplicates,nTargets] = size(Y);

    alpha = zeros(nTargets, nReplicates);
    beta = zeros(nTargets, nReplicates);
    gamma = zeros(1, nReplicates);
        
    for iE = 1:nReplicates
        for iT = 1:nTargets
            alpha(iT,iE) = P(idcs.alpha{iT}(iE));
            beta(iT,iE) = P(idcs.beta{iT}(iE));
        end
        gamma(iE) = P(idcs.gamma(iE));
        Y(:,iE,:) = circshift(Y(:,iE,:), gamma(iE),1);
    end
    
    y_mean = squeeze(nanmean(Y, 2));
    
    chi2_Y = zeros(nTargets, 1);
    for iT = 1:nTargets
        Y_tmp = squeeze(Y(:,:,iT));
        alpha_tmp = alpha(iT,:);
        beta_tmp = beta(iT,:);
        chi2_Y(iT) = ...
            nanmean(nansum( (Y_tmp - (alpha_tmp + beta_tmp.*y_mean(:,iT))).^2 ));
    end
    
    chi2 = mean(chi2_Y);
end